onnxruntime/core/providers/dml/DmlExecutionProvider/src/TensorDesc.cpp

```diff
@@ -315,3 +315,52 @@
     }
     m_bufferTensorDesc.DimensionCount = newDimensionCount;
 }
+
+// Uses dimensionMapping to reorder m_sizes and m_strides to match specific Tensor layout
+void TensorDesc::PermuteDimensions(const std::vector<uint32_t> dimensionMapping, const TensorAxis alignment)
+{
+    EnsureMinimumDimensionCount(dimensionMapping.size(), alignment);
+    InitializeStrides(static_cast<int>(dimensionMapping.size()), alignment);
+    PermuteArray(dimensionMapping, alignment);
+}
+
+// Shuffle m_sizes and m_strides acording to the indexes pointed by dimensionMapping
+void TensorDesc::PermuteArray(const std::vector<uint32_t> dimensionMapping, const TensorAxis alignment)
+{
+    std::vector<uint32_t> temp_sizes(dimensionMapping.size());
+    std::vector<uint32_t> temp_strides(dimensionMapping.size());
+    // Right alignment values are shifted to the end
+    int alignmentOffset = (alignment == TensorAxis::RightAligned) ? MaximumDimensionCount - static_cast<int>(dimensionMapping.size()) : 0;
+
+    for (size_t i = 0; i < dimensionMapping.size(); i++) {
+        temp_sizes[i] = m_sizes[dimensionMapping[i] + alignmentOffset];
+        temp_strides[i] = m_strides[dimensionMapping[i] + alignmentOffset];
+    }
+
+    std::copy(temp_sizes.begin(), temp_sizes.end(), m_sizes + alignmentOffset);
+    std::copy(temp_strides.begin(), temp_strides.end(), m_strides + alignmentOffset);
+    m_bufferTensorDesc.Sizes = m_sizes;
+    m_bufferTensorDesc.Strides = m_strides;
+}
+
+void TensorDesc::InitializeStrides(int dimensionCount, const TensorAxis alignment)
+{
+    ML_CHECK_VALID_ARGUMENT(alignment == TensorAxis::RightAligned || alignment == TensorAxis::LeftAligned);
+    // Right alignment values are shifted to the end
+    int alignmentOffset = (alignment == TensorAxis::RightAligned) ? MaximumDimensionCount - dimensionCount : 0;
+    int index = dimensionCount + alignmentOffset;
+    uint32_t stride = 1;
+    m_strides[index - 1] = 1;
+    for (int i = index - 2; i >= 0; i--)
+    {
+        stride *= m_sizes[i + 1];
+        m_strides[i] = stride;
+    }
+}
+
+void TensorDesc::EnsureMinimumDimensionCount(const size_t dimensionCount, const TensorAxis alignment)
```

Comment:
>const size_t dimensionCount [](http://example.com/codeflow?start=45&length=27)

`InitializeStrides(int dimensionCount...`

`EnsureMinimumDimensionCount(const size_t dimensionCount`

`SetDimensionCount(uint32_t newDimensionCount`

The dimension count is using 3 different types in 3 nearby functions. We should regularlize them consistently (uint32_t is the dominant type used for dimensions and follows `TensorDesc::m_minDimensionCount`)

Reply:
Its the way it gets used in each function, causes warnings which are treated as errors. I will give this another thought

Reply:
Yep, you will need to cast it at certain call sites.